303. Range Sum Query - Immutable
Given an integer array nums, handle multiple queries of the following type:

Calculate the sum of the elements of nums between indices left and right inclusive where left <= right.
Implement the NumArray class:

NumArray(int[] nums) Initializes the object with the integer array nums.
int sumRange(int left, int right) Returns the sum of the elements of nums between indices left and right inclusive (i.e. nums[left] + nums[left + 1] + ... + nums[right]).
 

Example 1:

Input
["NumArray", "sumRange", "sumRange", "sumRange"]
[[[-2, 0, 3, -5, 2, -1]], [0, 2], [2, 5], [0, 5]]
Output
[null, 1, -1, -3]

Explanation
NumArray numArray = new NumArray([-2, 0, 3, -5, 2, -1]);
numArray.sumRange(0, 2); // return (-2) + 0 + 3 = 1
numArray.sumRange(2, 5); // return 3 + (-5) + 2 + (-1) = -1
numArray.sumRange(0, 5); // return (-2) + 0 + 3 + (-5) + 2 + (-1) = -3
 

Constraints:

1 <= nums.length <= 104
-105 <= nums[i] <= 105
0 <= left <= right < nums.length
At most 104 calls will be made to sumRange.

In [1]:
from typing import List

class NumArray:

    def __init__(self, nums: List[int]):
        self.prefix = [0]*len(nums) # creating an array with 0 for the required size len(nums)
        self.prefix[0]=nums[0]
        for i in range (1,len(nums)):
            self.prefix[i]=self.prefix[i-1]+nums[i]


        

    def sumRange(self, left: int, right: int) -> int:
        # [-2, 0, 3, -5, 2, -1]
        # Prefix Array [-2,-2,1,-4,-2,-3]
        if left ==0:
            return self.prefix[right]
        else :
            return self.prefix[right] - self.prefix[left - 1]




nums = [-2, 0, 3, -5, 2, -1]
obj = NumArray(nums)
print(obj.sumRange(0,2))
print(obj.sumRange(2,5))
print(obj.sumRange(0,5))


# Time Complexity is O(n) --> creating the prefix Array
# O(1) --> getting the Sum

1
-1
-3


724. Find Pivot Index

Given an array of integers nums, calculate the pivot index of this array.

The pivot index is the index where the sum of all the numbers strictly to the left of the index is equal to the sum of all the numbers strictly to the index's right.

If the index is on the left edge of the array, then the left sum is 0 because there are no elements to the left. This also applies to the right edge of the array.

Return the leftmost pivot index. If no such index exists, return -1.

 

Example 1:

Input: nums = [1,7,3,6,5,6]
Output: 3
Explanation:
The pivot index is 3.
Left sum = nums[0] + nums[1] + nums[2] = 1 + 7 + 3 = 11
Right sum = nums[4] + nums[5] = 5 + 6 = 11
Example 2:

Input: nums = [1,2,3]
Output: -1
Explanation:
There is no index that satisfies the conditions in the problem statement.
Example 3:

Input: nums = [2,1,-1]
Output: 0
Explanation:
The pivot index is 0.
Left sum = 0 (no elements to the left of index 0)
Right sum = nums[1] + nums[2] = 1 + -1 = 0
 

Constraints:

1 <= nums.length <= 104
-1000 <= nums[i] <= 1000
 

Note: This question is the same as 1991: https://leetcode.com/problems/find-the-middle-index-in-array/

In [9]:
from typing import List

class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        if not nums:
            return -1
            
        # Create and populate the prefix sum array
        prefix = [0] * len(nums)
        prefix[0] = nums[0]
        for i in range(1, len(nums)):
            prefix[i] = prefix[i-1] + nums[i]
            
        total_sum = prefix[-1]
        
        # Check for the pivot index
        for i in range(0, len(nums)):
            # Handle left sum for the 0th index gracefully
            left_sum = prefix[i-1] if i > 0 else 0
            right_sum = total_sum - prefix[i]
            
            if left_sum == right_sum:
                print("pivot index = " + str(i))
                return i
                
        return -1

Sol = Solution()
Sol.pivotIndex([1,7,3,6,5,6])
Sol.pivotIndex([1,2,3])
Sol.pivotIndex( [2,1,-1])

pivot index = 3
pivot index = 0


0

Prefix Sum --> L to R Sum
Suffix Sum --> R to L Sum

if you want the fastest, cleanest execution in Python, you can use itertools.accumulate to handle the prefix sums natively. This is often faster in practice because the loop runs in optimized C under the hood.

In [17]:
from typing import List
import itertools

class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        total_sum = sum(nums)
        print(f"total is {total_sum}")
        
        # itertools.accumulate gives a running total: [nums[0], nums[0]+nums[1], ...]
        # Specifying initial=0 starts the accumulation at 0
        for i, left_sum in enumerate(itertools.accumulate(nums, initial=0)): # Returns (i=0,left_sum=1), (i=1,left_sum=3), (i=2,left_sum=6) 
            if i < len(nums) and left_sum == (total_sum - left_sum - nums[i]):
                print("pivot index = " + str(i))
                return i
        print("pivot index = -1")        
        return -1

Sol = Solution()
Sol.pivotIndex([1,7,3,6,5,6])
Sol.pivotIndex([1,2,3])
Sol.pivotIndex( [2,1,-1])

total is 28
pivot index = 3
total is 6
pivot index = -1
total is 2
pivot index = 0


0

In [18]:
from typing import List

class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        total_sum = sum(nums)
        left_sum = 0
        
        # Loop using standard indices, just like your code
        for i in range(len(nums)):
            right_sum = total_sum - left_sum - nums[i]
            
            if left_sum == right_sum:
                print("pivot index = " + str(i))
                return i
                
            # Update left_sum for the next iteration
            left_sum += nums[i]
            
        return -1
Sol = Solution()
Sol.pivotIndex([1,7,3,6,5,6])
Sol.pivotIndex([1,2,3])
Sol.pivotIndex( [2,1,-1])

pivot index = 3
pivot index = 0


0

In [19]:
# simpler way to create prefix Array
import itertools

# Your input list
nums = [1, 7, 3, 6, 5, 6]

# The command to create the prefix array
prefix = list(itertools.accumulate(nums))

print(prefix)
# Output: [1, 8, 11, 17, 22, 28]


[1, 8, 11, 17, 22, 28]


In [36]:
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prod_val = 1
        prod_tot_val = 1
        zero_counter = 0
        for i in range (0,len(nums)):
            if nums[i] != 0:
                prod_val *= nums[i]
            else :
                zero_counter += 1
            prod_tot_val *= nums[i]
        res = [0]*len(nums)
        for i in range (0,len(nums)):
            if zero_counter >= len(nums)-1:
                res = [0]*len(nums)
            elif nums[i]==0:
                res[i]=prod_val
            else:
                res[i]=int(prod_tot_val/nums[i])
        print(res)
        return res
            

Sol = Solution()
Sol.productExceptSelf([1,2,3,4])
Sol.productExceptSelf([24,12,8,6]) 
Sol.productExceptSelf([-1,1,0,-3,3])
Sol.productExceptSelf([0,0])
Sol.productExceptSelf([0,4,0])

[24, 12, 8, 6]
[576, 1152, 1728, 2304]
[0, 0, 9, 0, 0]
[0, 0]
[0, 0, 0]


[0, 0, 0]

In [ ]:
import itertools
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prod_val = 1
        prod_tot_val = 1
        zero_counter = 0
        for i in range (0,len(nums)):
            if nums[i] != 0:
                prod_val *= nums[i]
            else :
                zero_counter += 1
            prod_tot_val *= nums[i]
        res = [0]*len(nums)
        for i in range (0,len(nums)):
            if zero_counter >= len(nums)-1:
                res = [0]*len(nums)
            elif nums[i]==0:
                res[i]=prod_val
            else:
                res[i]=int(prod_tot_val/nums[i])
        print(res)
        # res1 = [0]*len(nums)
        # Pre = [0]*len(nums)
        # suf = [0]*len(nums)
        # pref = list(itertools.accumulate(nums))
        # nums.reverse()
        # sufi = list(itertools.accumulate(nums))
        res = [pref[i] + sufi[i] for i in range(len(pref))]
        print (pref , sufi)
        print (res)
        return res
            

Sol = Solution()
Sol.productExceptSelf([1,2,3,4])
Sol.productExceptSelf([24,12,8,6]) 
Sol.productExceptSelf([-1,1,0,-3,3])
Sol.productExceptSelf([0,0])
Sol.productExceptSelf([0,4,0])

[24, 12, 8, 6]
[1, 3, 6, 10] [4, 7, 9, 10]
[5, 10, 15, 20]
[576, 1152, 1728, 2304]
[24, 36, 44, 50] [6, 14, 26, 50]
[30, 50, 70, 100]
[0, 0, 9, 0, 0]
[-1, 0, 0, -3, 0] [3, 0, 0, 1, 0]
[2, 0, 0, -2, 0]
[0, 0]
[0, 0] [0, 0]
[0, 0]
[0, 0, 0]
[0, 4, 4] [0, 4, 4]
[0, 8, 8]


[0, 8, 8]

In [1]:
# from typing import List

# class Solution:
#     def productExceptSelf(self, nums: List[int]) -> List[int]:
#         prod_val = 1
#         prod_tot_val = 1
#         zero_counter = 0
#         for i in range (0,len(nums)):
#             if nums[i] != 0:
#                 prod_val *= nums[i]
#             else :
#                 zero_counter += 1
#             prod_tot_val *= nums[i]
#         res = [0]*len(nums)
#         for i in range (0,len(nums)):
#             if zero_counter >= len(nums)-1:
#                 res = [0]*len(nums)
#             elif nums[i]==0:
#                 res[i]=prod_val
#             else:
#                 res[i]=int(prod_tot_val/nums[i])
#         # print(res)
#         return res

# Sol = Solution()
# Sol.productExceptSelf([1,2,3,4])
# Sol.productExceptSelf([24,12,8,6]) 
# Sol.productExceptSelf([-1,1,0,-3,3])
# Sol.productExceptSelf([0,0])
# Sol.productExceptSelf([0,4,0])

[0, 0, 0]

2 pointer Array

Str Reversal

In [14]:
class two_p:
    def validate_palindrome(self, s: str):
        l=0
        r= len(s)-1
        print ( "checking string "+s)
        while l<=r :
            if s[l] == s[r]:
                pass
            else :
                print("not a palindrome")
                return
            l+=1
            r-=1
        print("is a palindrome")

obj = two_p()
obj.validate_palindrome("abcba")
obj.validate_palindrome("abcb")
obj.validate_palindrome("abccba")


checking string abcba
is a palindrome
checking string abcb
not a palindrome
checking string abccba
is a palindrome


Given a sorted array of integer and a target, retuen true/false if there exists a pair that sum upto the target
[1,3,5,9,10]

target = 8

are there any 2 elements that I can add to get the value of target ie 8?
ans: 3,5

output: true

if target = 2?
output: False

[1,3,5,9,11]
t=8
l=1, r=11

if l+r=t return l,r
elif t<l+r -> r-1 -> r=9
elif t>l+r -> l+1 -> l=3


This works because array is sorted.

In [24]:
class targ:
    def get_val(self, nums: list, t: int):
        print(f"checking for {nums} and {t}")
        l=0
        r=len(nums)-1
        while (l<r):
            sum=nums[l]+nums[r]
            if t==sum:
                print (f" {nums[l]} and {nums[r]} can be added to get {t}")
                return
            elif t<sum:
                r-=1
            else: 
                l+=1
        print(f"no 2 values can be added to get {t}")

obj=targ()
obj.get_val([1,2,5,9,10],8)
obj.get_val([1,2,5,9,10],7)
obj.get_val([1,3,5,9,10],8)
obj.get_val([1,2,5,6,12,11],11)
obj.get_val([1],11)


checking for [1, 2, 5, 9, 10] and 8
no 2 values can be added to get 8
checking for [1, 2, 5, 9, 10] and 7
 2 and 5 can be added to get 7
checking for [1, 3, 5, 9, 10] and 8
 3 and 5 can be added to get 8
checking for [1, 2, 5, 6, 12, 11] and 11
 5 and 6 can be added to get 11
checking for [1] and 11
no 2 values can be added to get 11


If the array was not sorted?

In [28]:
class targ:
    def get_val(self, nums: list, t: int):
        hashmap = {}
        for i, num in enumerate(nums):
            complement = t - num

            if complement in hashmap:
                print(f" {complement} and {num} can be added to get {t}")
                return

            hashmap[num] = i

        print(f"no 2 values can be added to get {t}")
        

obj=targ()
obj.get_val([12,1,2,5,9,10],8)
obj.get_val([12,1,2,5,9,10],21)
obj.get_val([12,1,3,5,9,10],8)
obj.get_val([12,1,2,5,6,12,11],11)
obj.get_val([1],11)


no 2 values can be added to get 8
 12 and 9 can be added to get 21
 3 and 5 can be added to get 8
 5 and 6 can be added to get 11
no 2 values can be added to get 11


Checking if array is sorted

In [30]:
class sorted:
    def check(self,nums:list):
        print(f"checking for {nums}")
        l=0
        r=1
        while r<len(nums):
            if nums[l]>nums[r]:
                print("array is not sorted")
                return
            l+=1
            r+=1
        print("array is sorted")

obj = sorted()
obj.check([1,2,3])
obj.check([1,2,5,3]) 


checking for [1, 2, 3]
array is sorted
checking for [1, 2, 5, 3]
array is not sorted


Leetcode 15: 3 Sum  **Important**
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.

 

Example 1:

Input: nums = [-1,0,1,2,-1,-4]
Output: [[-1,-1,2],[-1,0,1]]
Explanation: 
nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
The distinct triplets are [-1,0,1] and [-1,-1,2].
Notice that the order of the output and the order of the triplets does not matter.
Example 2:

Input: nums = [0,1,1]
Output: []
Explanation: The only possible triplet does not sum up to 0.
Example 3:

Input: nums = [0,0,0]
Output: [[0,0,0]]
Explanation: The only possible triplet sums up to 0.
 

Constraints:

3 <= nums.length <= 3000
-105 <= nums[i] <= 105

In [ ]:
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        res=[]
        for i in range(len(nums)-2):
            left = i+1
            right = len(nums)-1
            while left<right:
                total=nums[i]+nums[left]+nums[right]
                if total == 0:
                    triplet = [nums[i], nums[left], nums[right]]
                    if triplet not in res:
                        res.append(triplet)
                    left+=1
                    right-=1

                    while left < right and nums[left] == nums[left-1]:
                        left+=1
                    while left<right and nums[right]== nums[right+1]:
                        right-=1
                elif total>0:
                    right-=1
                else: 
                    left+=1
        return res

    
# complexity:
# Sort -> n log n
# nested while loop -> n^2
# so total complexity is n log (n) + n^2




obj = Solution()
print(obj.threeSum([-2,-1,-1,0,0,1]))
print(obj.threeSum([-1,0,1,2,-1,-4]))

        

[[-1, 0, 1]]
[[-1, -1, 2], [-1, 0, 1]]


complexity:
Sort -> n log n
nested while loop -> n^2

so total complexity is n log (n) + n^2

Leet code 
1. Two Sum

You are given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 

Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
Example 2:

Input: nums = [3,2,4], target = 6
Output: [1,2]
Example 3:

Input: nums = [3,3], target = 6
Output: [0,1]
 

Constraints:

2 <= nums.length <= 104
-109 <= nums[i] <= 109
-109 <= target <= 109
Only one valid answer exists.
 

Follow-up: Can you come up with an algorithm that is less than O(n2) time complexity?

In [10]:
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        hashmap = {}

        for i, num in enumerate(nums):
            complement = target - num

            if complement in hashmap:
                return [hashmap[complement], i]

            hashmap[num] = i
            print (hashmap[num])

        return []

obj = Solution()
obj.twoSum([1, 2, 3, 4], 6)

0
1
2


[1, 3]

Q - Count freequency of each element

In [ ]:
class Solution:
    def freq(self, nums: list[int]):
        hashmap = {}
        print (enumerate(nums))
        for i, num in enumerate(nums):
            if num in hashmap:
                hashmap[num] += 1
            else:
                hashmap[num] = 1
        print (hashmap)

obj = Solution()
obj.freq([1, 2, 3, 4, 2, 3])

{1: 1, 2: 2, 3: 2, 4: 1}


169: Majority Element
Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.

 

Example 1:

Input: nums = [3,2,3]
Output: 3
Example 2:

Input: nums = [2,2,1,1,1,2,2]
Output: 2
 

Constraints:

n == nums.length
1 <= n <= 5 * 104
-109 <= nums[i] <= 109
The input is generated such that a majority element will exist in the array.
 

Follow-up: Could you solve the problem in linear time and in O(1) space?

In [ ]:
class Solution:
    def majorityElement(self, nums: list[int]) -> int:
        hashmap={}
        thr = len(nums)/2
        for i in nums:
            if i in hashmap:
                hashmap[i]+=1
            else:
                hashmap[i] = 1
            if hashmap[i] > thr :
                print(hashmap[i])
                return i
                
            # print(hashmap)
            

obj = Solution()
obj.majorityElement([1,1,1,1,2,2,2,2,2,2,3])
        

6


2

given a string s count the character that is coming max no of times



abbaacdee
op:a

In [ ]:
class Solution:
    def majorityElement(self, st: str):
        hashmap={}
        max_val = "A"
        max_count = 0
        for i in st:
            if i in hashmap:
                hashmap[i]+=1
            else:
                hashmap[i] = 1
            if hashmap[i] > max_count:
                max_val = i
                max_count = hashmap[i]
        print (max_count, max_val)
                
            # print(hashmap)
            

obj = Solution()
obj.majorityElement("abcabcaabc")
        

4 a


Sliding Window

[2,1,5,1,3,2]
Find the max sum of 3 consecutive elements

The idea is remove one element and add a new element and slide in this way



In [8]:
def slide (nums: list[int]):
    sum = nums[0]+nums[1]+nums[2]
    max = sum
    for i in range(len(nums)-3):
        sum -= nums[i]
        sum += nums[i+3]
        if (sum > max):
            max = sum
        # print (sum, max)
    print (max)

slide ([2,1,5,1,3,2])


9


Get max avg of continuous k elements

In [3]:
def findMaxAverage( nums: list[int], k: int) -> float:
        sum = 0
        for i in range (k):
            sum +=nums[i]
        avg = sum/k

        for i in range (len(nums)-k):
            sum -= nums[i]
            sum += nums[i+k]
            new_avg = sum/k
            if (avg > new_avg):
                avg = new_avg
        return avg

findMaxAverage ([1,12,-5,-6,50,3],4)

0.5

560. Subarray Sum Equals K

Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.

A subarray is a contiguous non-empty sequence of elements within an array.

 

Example 1:

Input: nums = [1,1,1], k = 2
Output: 2
Example 2:

Input: nums = [1,2,3], k = 3
Output: 2
 

Constraints:

1 <= nums.length <= 2 * 104
-1000 <= nums[i] <= 1000
-107 <= k <= 107

In [10]:
from typing import List
from collections import defaultdict

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        count = 0
        prefix_sum = 0
        freq = defaultdict(int)
        freq[0] = 1  

        for num in nums:
            prefix_sum += num
            count += freq[prefix_sum - k]
            freq[prefix_sum] += 1
            print (num,prefix_sum,k,prefix_sum - k, count,freq[prefix_sum - k],freq[prefix_sum] ,freq)

        return count

obj=Solution()
obj.subarraySum([28,54,7,-70,22,65,-6],100)
obj.subarraySum([28,54,7,-70,22,65,-6],87)

28 28 100 -72 0 0 1 defaultdict(<class 'int'>, {0: 1, -72: 0, 28: 1})
54 82 100 -18 0 0 1 defaultdict(<class 'int'>, {0: 1, -72: 0, 28: 1, -18: 0, 82: 1})
7 89 100 -11 0 0 1 defaultdict(<class 'int'>, {0: 1, -72: 0, 28: 1, -18: 0, 82: 1, -11: 0, 89: 1})
-70 19 100 -81 0 0 1 defaultdict(<class 'int'>, {0: 1, -72: 0, 28: 1, -18: 0, 82: 1, -11: 0, 89: 1, -81: 0, 19: 1})
22 41 100 -59 0 0 1 defaultdict(<class 'int'>, {0: 1, -72: 0, 28: 1, -18: 0, 82: 1, -11: 0, 89: 1, -81: 0, 19: 1, -59: 0, 41: 1})
65 106 100 6 0 0 1 defaultdict(<class 'int'>, {0: 1, -72: 0, 28: 1, -18: 0, 82: 1, -11: 0, 89: 1, -81: 0, 19: 1, -59: 0, 41: 1, 6: 0, 106: 1})
-6 100 100 0 1 1 1 defaultdict(<class 'int'>, {0: 1, -72: 0, 28: 1, -18: 0, 82: 1, -11: 0, 89: 1, -81: 0, 19: 1, -59: 0, 41: 1, 6: 0, 106: 1, 100: 1})
28 28 87 -59 0 0 1 defaultdict(<class 'int'>, {0: 1, -59: 0, 28: 1})
54 82 87 -5 0 0 1 defaultdict(<class 'int'>, {0: 1, -59: 0, 28: 1, -5: 0, 82: 1})
7 89 87 2 0 0 1 defaultdict(<class 'int'>, {0: 1, -59: 

1

Here’s the “aha moment” you’re looking for — why only the *right* prefix sum value shows up twice and triggers the count.

---

### Visual Timeline of Prefix Sums
Think of prefix sums as points on a number line. Each step forward adds the next number.

For your case:  
`nums = [28,54,7,-70,22,65,-6], k = 87`

Prefix sums evolve like this:

```
Index:   -1    0    1    2     3     4     5     6
Value:    0   28   82   89    19    41   106   100
```

---

### The Key Overlap
At index 3, prefix sum = **19**.  
At index 5, prefix sum = **106**.  

Now check:  
`106 - 87 = 19` → bingo!  

That means the subarray from index 4 to 5 (`[22,65]`) sums to 87.

---

### Why Only the “Right” Value Comes Twice
- Every time you advance, you add the current number to the running sum.  
- The hashmap stores **all past prefix sums**.  
- When you hit a new prefix sum, you check if subtracting `k` lands on a past prefix sum.  
- If yes, that overlap means the slice between those two indices is exactly `k`.

So in this case:
- The “left” prefix sum (19) was stored earlier.  
- The “right” prefix sum (106) is the current one.  
- Their difference is exactly 87, so the algorithm knows `[22,65]` is valid.

---

### Visual Diagram (mental picture)
Imagine prefix sums as dots on a line:

```
0 ----28----82----89----19----41----106----100
```

Draw an arrow from `19` (earlier) to `106` (later).  
The gap between them is exactly `87`.  
That arrow corresponds to the subarray `[22,65]`.

---

✨ The “aha” is: **the algorithm doesn’t store subarrays — it stores prefix sums. When a later prefix sum minus `k` equals an earlier prefix sum, the difference between those two points is the subarray sum.**  

Would you like me to actually sketch this timeline with arrows in a diagram so you can *see* the overlap visually?

BInary Search of a sorted array

TIME COMPLEXITY: 
checking values of n, n/2,n/4,... until element is found
so complexity is O(log n)


In [14]:
from typing import List

class Solution:
    def binary_search(self, nums: List[int], k: int):
        l = 0
        r = len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            value = nums[mid]
            if nums[l] == k:
                print(nums[l], l)
                return
            elif nums[r] == k:
                print(nums[r], r)
                return            
            elif value == k:
                print(value, mid)
                return
            elif value < k:
                l = mid + 1
            else:
                r = mid - 1

        print("element not found")
        return

obj = Solution()
obj.binary_search([1, 2, 3, 4, 5, 6], 5)


5 4


34. Find First and Last Position of Element in Sorted Array

Given an array of integers nums sorted in non-decreasing order, find the starting and ending position of a given target value.

If target is not found in the array, return [-1, -1].

You must write an algorithm with O(log n) runtime complexity.

 

Example 1:

Input: nums = [5,7,7,8,8,10], target = 8
Output: [3,4]
Example 2:

Input: nums = [5,7,7,8,8,10], target = 6
Output: [-1,-1]
Example 3:

Input: nums = [], target = 0
Output: [-1,-1]
 

Constraints:

0 <= nums.length <= 105
-109 <= nums[i] <= 109
nums is a non-decreasing array.
-109 <= target <= 109

In [23]:
from typing import List

class Solution:
    def searchRange(self, nums: List[int], k: int) -> List[int]:
        def boundSearch(nums, k, findLeft: bool) -> int:
            l, r = 0, len(nums) - 1
            ans = -1
            while l <= r:
                mid = (l + r) // 2
                if nums[mid] == k:
                    ans = mid
                    if findLeft:
                        r = mid - 1   # keep searching left
                    else:
                        l = mid + 1   # keep searching right
                elif nums[mid] < k:
                    l = mid + 1
                else:
                    r = mid - 1
            return ans

        start = boundSearch(nums, k, True)
        end = boundSearch(nums, k, False)
        return [start, end]


obj = Solution()
print(obj.searchRange([1, 2, 3, 3, 3, 6], 3))


[2, 4]


In [ ]:
33. Search in Rotated Sorted Array

There is an integer array nums sorted in ascending order (with distinct values).

Prior to being passed to your function, nums is possibly left rotated at an unknown index k (1 <= k < nums.length) such that the resulting array is [nums[k], nums[k+1], ..., nums[n-1], nums[0], nums[1], ..., nums[k-1]] (0-indexed). For example, [0,1,2,4,5,6,7] might be left rotated by 3 indices and become [4,5,6,7,0,1,2].

Given the array nums after the possible rotation and an integer target, return the index of target if it is in nums, or -1 if it is not in nums.

You must write an algorithm with O(log n) runtime complexity.

 

Example 1:

Input: nums = [4,5,6,7,0,1,2], target = 0
Output: 4
Example 2:

Input: nums = [4,5,6,7,0,1,2], target = 3
Output: -1
Example 3:

Input: nums = [1], target = 0
Output: -1
 

Constraints:

1 <= nums.length <= 5000
-104 <= nums[i] <= 104
All values of nums are unique.
nums is an ascending array that is possibly rotated.
-104 <= target <= 104


In [50]:
from typing import List

class Solution:
    def search(self, nums: List[int], k: int):
        l = 0
        r = len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            value = nums[mid]
            # print(f"start {value, k, nums[l],nums[r]}")
            if nums[l] == k:
                print(nums[l], l)
                return
            elif nums[r] == k:
                print(nums[r], r)
                return            
            elif value == k:
                print(value, mid)
                return
            elif value > nums[l] :
                    if (value > k > nums[l]):
                        r = mid - 1
                    else:
                        l = mid + 1
            else:
                    if (value < k < nums[r]):
                        l = mid + 1
                    else:
                        r = mid - 1
                # print(f"r = mid - 1 {r,mid}")
            
                # print(f"l=mid+1 {l,mid}")
                
            # print(value, k, nums[l],nums[r])

        print("element not found")
        return

obj = Solution()
obj.search([4, 5, 6, 1, 2], 5)
obj.search([4, 5, 6, 1, 2], 3)
obj.search([4, 5, 6, 1, 2], 7)
obj.search([4, 5, 6, 7, 2], 7)
obj.search([4,5,6,7,0,1,2], 0)
obj.search([5,1,2,3,4],1)


5 1
element not found
element not found
7 3
0 4
1 1


Binary search II
Binary search on answer, not searching an element, we are searching for the min, max values that satisfy the condition
Monotonic --> once the anwer changes from one state to another, it doesnt change back
can you reach 50km in 1 hr at 10kmph? no,
20kmph? no
...
40kmph? no
50kmph? yes--> monotonic increasing --> any speed above this will remain yes
60kmph? yes


monotonic decreasing --> any speed below this will take more than one hour

Clue: *minimum* time, *maximum* speed, *minimum* capacity, etc 

875. Koko Eating Bananas
Koko loves to eat bananas. There are n piles of bananas, the ith pile has piles[i] bananas. The guards have gone and will come back in h hours.

Koko can decide her bananas-per-hour eating speed of k. Each hour, she chooses some pile of bananas and eats k bananas from that pile. If the pile has less than k bananas, she eats all of them instead and will not eat any more bananas during this hour.

Koko likes to eat slowly but still wants to finish eating all the bananas before the guards return.

Return the minimum integer k such that she can eat all the bananas within h hours.

 

Example 1:

Input: piles = [3,6,7,11], h = 8
Output: 4
Example 2:

Input: piles = [30,11,23,4,20], h = 5
Output: 30
Example 3:

Input: piles = [30,11,23,4,20], h = 6
Output: 23
 

Constraints:

1 <= piles.length <= 104
piles.length <= h <= 109
1 <= piles[i] <= 109

In [ ]:
class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        
        def ispossible(mid):
            hours=0
            for pile in piles:
                hours+=(pile+mid-1)//mid
            return hours <= h
        
        low = 1
        high = max(piles)
        answer = high
        while low <= high:
            mid = (low+high)//2
            if ispossible(mid):
                answer = mid
                high = mid-1
            else:
                low=mid+1
        return answer
        

Stacks: Last In First Out LIFO like a stack of plates in a wedding hall, the last plate kept on the stack will be the first one to be picked


Fundamentals:
Push: add to top
Pop: remove frmo top
Peek: ONLY RETURN top element but DO NOT Remove it like pop would
isEmpty:

In [2]:
stack = []
stack.append(10)
stack.append(20)
stack.append(30) # Push

print (stack[-1]) #peek - return the top of stack WITHOUT REMOVING IT
print(stack.pop())
print(stack.pop()) #removing the element



30
30
20


20. Valid Parentheses

Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.
 

Example 1:

Input: s = "()"

Output: true

Example 2:

Input: s = "()[]{}"

Output: true

Example 3:

Input: s = "(]"

Output: false

Example 4:

Input: s = "([])"

Output: true

Example 5:

Input: s = "([)]"

Output: false

 

Constraints:

1 <= s.length <= 104
s consists of parentheses only '()[]{}'.

In [ ]:
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {")": "(", "]": "[", "}": "{"}
        
        for char in s:
            if char in mapping.values():  # opening bracket
                stack.append(char)
            elif char in mapping:  # closing bracket
                if not stack or stack[-1] != mapping[char]: # check if stack is empty or if the closing bracket is != opening one
                    return False
                stack.pop()
        
        return not stack


# Test
obj = Solution()
print(obj.isValid("{}()[]"))   # ✅ True
print(obj.isValid("{[}]"))     # ❌ False
print(obj.isValid("((()))"))   # ✅ True
print(obj.isValid("(()"))      # ❌ False


True
False
True
False


Another way to solve the above problem

In [8]:
class Solution:
    def isValid(self, s: str) -> bool:
        check = []  # use list as stack
        for a in s:
            if a in ("{", "(", "["):
                check.append(a)
            elif a == ")":
                if not check or check[-1] != "(":
                    return False
                check.pop()
            elif a == "]":
                if not check or check[-1] != "[":
                    return False
                check.pop()
            elif a == "}":
                if not check or check[-1] != "{":
                    return False
                check.pop()
        return not check  # stack should be empty if valid


# Test
obj = Solution()
print(obj.isValid("{}()[]"))   # ✅ True
print(obj.isValid("{[}]"))     # ❌ False
print(obj.isValid("((()))"))   # ✅ True
print(obj.isValid("(()"))      # ❌ False


True
False
True
False


In [10]:
class Solution:
    def isValid(self, s: str) -> bool:
        check = []  # use list as stack
        for a in s:
            if a in ("{", "(", "["):
                check.append(a)
            elif a in ("}", ")", "]"):
                top = check[-1]
                if( a == ")" and top != "(" or a=='}' and top != "{" or a==']' and top!= "["):
                    return False
            check.pop()
        return not check  # stack should be empty if valid


# Test
obj = Solution()
print(obj.isValid("{}()[]"))   # ✅ True
print(obj.isValid("{[}]"))     # ❌ False
print(obj.isValid("((()))"))   # ✅ True
print(obj.isValid("(()"))      # ❌ False

IndexError: list index out of range

1047. Remove All Adjacent Duplicates In String

You are given a string s consisting of lowercase English letters. A duplicate removal consists of choosing two adjacent and equal letters and removing them.

We repeatedly make duplicate removals on s until we no longer can.

Return the final string after all such duplicate removals have been made. It can be proven that the answer is unique.

 

Example 1:

Input: s = "abbaca"
Output: "ca"
Explanation: 
For example, in "abbaca" we could remove "bb" since the letters are adjacent and equal, and this is the only possible move.  The result of this move is that the string is "aaca", of which only "aa" is possible, so the final string is "ca".
Example 2:

Input: s = "azxxzy"
Output: "ay"
 

Constraints:

1 <= s.length <= 105
s consists of lowercase English letters.

In [16]:
class Solution:
    def removeDuplicates(self, s: str) -> str:
        stack=[]
        for a in s:
            if len(stack) == 0:
                stack.append(a)
            elif a == stack[-1]:
                stack.pop()
            else:
                stack.append(a)
        res = ''
        while len(stack)!=0:
            res+=stack[-1]
            stack.pop()
        return res[::-1]

obj = Solution()
obj.removeDuplicates("abbaya")

'ya'

better way to do it below

In [ ]:
class Solution:
    def removeDuplicates(self, s: str) -> str:
        stack = []
        for a in s:
            if stack and stack[-1] == a:
                stack.pop()
            else:
                stack.append(a)
        return "".join(stack)   # directly join the stack into a string


obj = Solution()
print(obj.removeDuplicates("abbaya"))  # Output: "ya"


ya


get the next highest element:

In [ ]:
class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        n = len(temperatures)
        output = [0] * n
        stack = []  # will store indices
        for i in range(n):
            if len(stack)!=0 and temperatures[i]>temperatures[stack[-1]]:
                j=stack.pop()
                output[j]=temperatures[i]
            else:
                stack.append(i)
        return output
        # [74, 0, 0, 76, 72, 0, 0, 0]


obj = Solution()
print(obj.dailyTemperatures([73,74,75,71,69,72,76,73]))


739. Daily Temperatures
Given an array of integers temperatures represents the daily temperatures, return an array answer such that answer[i] is the number of days you have to wait after the ith day to get a warmer temperature. If there is no future day for which this is possible, keep answer[i] == 0 instead.

 

Example 1:

Input: temperatures = [73,74,75,71,69,72,76,73]
Output: [1,1,4,2,1,1,0,0]
Example 2:

Input: temperatures = [30,40,50,60]
Output: [1,1,1,0]
Example 3:

Input: temperatures = [30,60,90]
Output: [1,1,0]
 

Constraints:

1 <= temperatures.length <= 105
30 <= temperatures[i] <= 100

Monotonic Stack: 

In [ ]:
class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        n = len(temperatures)
        output = [0] * n
        stack = []  # will store indices    

        # This will also work, copilot version
        # for i, temp in enumerate(temperatures):
        #     # while current temp is greater than the last stacked temp
        #     while stack and temp > temperatures[stack[-1]]:
        #         prev_index = stack.pop()
        #         output[prev_index] = i - prev_index
        #     stack.append(i)

        for i in range (n):
            print(f"=============iteration {i}=============")
            while len(stack)!=0 and temperatures[i]>temperatures[stack[-1]]:
                print("----------next while------")
                print(f"temperatures[i]>temperatures[stack[-1]]{temperatures[i],temperatures[stack[-1]]}")
                print(f"stack {stack},output {output}")
                j=stack.pop() 
                # Loop in while until all the elements satistfying > condition are satisfied
                output[j]=i-j 
                #Use output[j] because you have added all the elements 
                #that are lesser than in the else block
                print(f"j {j},output {output}")
                print("-----------end while---------")
            else:
                print("----------next else------")
                print(f"else: Adding {temperatures[i],i} to stack")
                stack.append(i)
                print(f"stack {stack}")
                print("-----------end else---------")
                # if: [1, 0, 0, 3, 1, 0, 0, 0]


        return output


obj = Solution()
print(obj.dailyTemperatures([73,74,75,71,69,72,76,73]))


=============iteration 0=============
----------next else------
else: Adding (73, 0) to stack
stack [0]
-----------end else---------
=============iteration 1=============
----------next while------
temperatures[i]>temperatures[stack[-1]](74, 73)
stack [0],output [0, 0, 0, 0, 0, 0, 0, 0]
j 0,output [1, 0, 0, 0, 0, 0, 0, 0]
-----------end while---------
----------next else------
else: Adding (74, 1) to stack
stack [1]
-----------end else---------
=============iteration 2=============
----------next while------
temperatures[i]>temperatures[stack[-1]](75, 74)
stack [1],output [1, 0, 0, 0, 0, 0, 0, 0]
j 1,output [1, 1, 0, 0, 0, 0, 0, 0]
-----------end while---------
----------next else------
else: Adding (75, 2) to stack
stack [2]
-----------end else---------
=============iteration 3=============
----------next else------
else: Adding (71, 3) to stack
stack [2, 3]
-----------end else---------
=============iteration 4=============
----------next else------
else: Adding (69, 4) to stack
stack

Queue : FIFO
Add to the end -> Enqueue -> append(x)
Remove from first -> Dequeue -> popleft()
Look at the first element in the first -> Peek/front
isEmpty


X LIST X -> q.pop() -> Remove the element will lead to the elements being shifted --> This is why we cant use "List" for Queue -> complexity O(n)








In [1]:
from collections import deque

q = deque()
q.append(10)
q.append(20)
q.append(30)
print (q.popleft())
print(q[0]) # Peek


10
20


Window : [2,4,6,8,10]
fixed size window: 3 element window
variable/time size/based window: data in the last 3000 ms 
Queue: Old data will expire and new data will enter 
keep only relevant data



Moving avg from datastream


Moving Average from Data Stream
Given a stream of integers and a window size, calculate the moving average of all integers in the sliding window.

Implement the MovingAverage class:

MovingAverage(int size) Initializes the object with the size of the window size.
double next(int val) Returns the moving average of the last size values of the stream.
Example 1:

Input
["MovingAverage", "next", "next", "next", "next"]
[[3], [1], [10], [3], [5]]
Output
[null, 1.0, 5.5, 4.66667, 6.0]
Explanation
MovingAverage movingAverage = new MovingAverage(3);
movingAverage.next(1); // return 1.0 = 1 / 1
movingAverage.next(10); // return 5.5 = (1 + 10) / 2
movingAverage.next(3); // return 4.66667 = (1 + 10 + 3) / 3
movingAverage.next(5); // return 6.0 = (10 + 3 + 5) / 3
Constraints:

1 <= size <= 1000
-105 <= val <= 105
At most 104 calls will be made to next.
Solution
Approach 1: Array or List
Intuition

Following the description of the problem, we could simply keep track of all the incoming values with the data structure of Array or List. Then from the data structure, later we retrieve the necessary elements to calculate the average.

Press enter or click to view image in full size

Algorithm

First, we initialize a variable queue to store the values from the data stream, and the variable n for the size of the moving window.
At each invocation of next(val), we first append the value to the queue. We then retrieve the last n values from the queue, in order to calculate the average.
class MovingAverage {
  int size;
  List queue = new ArrayList<Integer>();
  public MovingAverage(int size) {
    this.size = size;
  }
public double next(int val) {
    queue.add(val);
    // calculate the sum of the moving window
    int windowSum = 0;
    for(int i = Math.max(0, queue.size() - size); i < queue.size(); ++i)
      windowSum += (int)queue.get(i);
return windowSum * 1.0 / Math.min(queue.size(), size);
  }
}
Complexity

Time Complexity: \mathcal{O}(N)O(N) where NN is the size of the moving window, since we need to retrieve NN elements from the queue at each invocation of next(val) function.
Space Complexity: \mathcal{O}(M)O(M), where MM is the length of the queue which would grow at each invocation of the next(val) function.
Approach 2: Double-ended Queue
Intuition

Get Santanu Das’s stories in your inbox
Join Medium for free to get updates from this writer.

Enter your email
Subscribe

Remember me for faster sign in

We could do better than the first approach in both time and space complexity.

First of all, one might notice that we do not need to keep all values from the data stream, but rather the last n values which falls into the moving window.

By definition of the moving window, at each step, we add a new element to the window, and at the same time we remove the oldest element from the window. Here, we could apply a data structure called double-ended queue (a.k.a deque) to implement the moving window, which would have the constant time complexity (\mathcal{O}(1)O(1)) to add or remove an element from both its ends. With the deque, we could reduce the space complexity down to \mathcal{O}(N)O(N) where NN is the size of the moving window.

Press enter or click to view image in full size

Secondly, to calculate the sum, we do not need to reiterate the elements in the moving window.

We could keep the sum of the previous moving window, then in order to obtain the sum of the new moving window, we simply add the new element and deduce the oldest element. With this measure, we then can reduce the time complexity to constant.

Algorithm

Here is the definition of the deque from Python. We have similar implementation of deque in other programming languages such as Java.

Deques are a generalization of stacks and queues (the name is pronounced deck and is short for double-ended queue). Deques support thread-safe, memory efficient appends and pops from either side of the deque with approximately the same O(1) performance in either direction.

Follow the intuition, we replace the queue with the deque and add a new variable window_sum in order to calculate the sum of moving window in constant time.

class MovingAverage {
  int size, windowSum = 0, count = 0;
  Deque queue = new ArrayDeque<Integer>();
public MovingAverage(int size) {
    this.size = size;
  }
public double next(int val) {
    ++count;
    // calculate the new sum by shifting the window
    queue.add(val);
    int tail = count > size ? (int)queue.poll() : 0;
windowSum = windowSum - tail + val;
return windowSum * 1.0 / Math.min(size, count);
  }
}
Complexity

Time Complexity: \mathcal{O}(1)O(1), as we explained in intuition.
Space Complexity: \mathcal{O}(N)O(N), where NN is the size of the moving window.
Approach 3: Circular Queue with Array
Intuition

Other than the deque data structure, one could also apply another fun data structure called circular queue, which is basically a queue with the circular shape.

Press enter or click to view image in full size

The major advantage of circular queue is that by adding a new element to a full circular queue, it automatically discards the oldest element. Unlike deque, we do not need to explicitly remove the oldest element.
Another advantage of circular queue is that a single index suffices to keep track of both ends of the queue, unlike deque where we have to keep a pointer for each end.
Algorithm

No need to resort to any library, one could easily implement a circular queue with a fixed-size array. The key to the implementation is the correlation between the index of head and tail elements, which we could summarize in the following formula:

\text{tail} = (\text{head} + 1) \mod \text{size}tail=(head+1)modsize

In other words, the tail element is right next to the head element. Once we move the head forward, we would overwrite the previous tail element.

class MovingAverage {
  int size, head = 0, windowSum = 0, count = 0;
  int[] queue;
  public MovingAverage(int size) {
    this.size = size;
    queue = new int[size];
  }
public double next(int val) {
    ++count;
    // calculate the new sum by shifting the window
    int tail = (head + 1) % size;
    windowSum = windowSum - queue[tail] + val;
    // move on to the next head
    head = (head + 1) % size;
    queue[head] = val;
    return windowSum * 1.0 / Math.min(size, count);
  }
}

Input
["MovingAverage", "next", "next", "next", "next"]
[[3], [1], [10], [3], [5]]
Output
[null, 1.0, 5.5, 4.66667, 6.0]
Explanation
MovingAverage movingAverage = new MovingAverage(3);
movingAverage.next(1); // return 1.0 = 1 / 1
movingAverage.next(10); // return 5.5 = (1 + 10) / 2
movingAverage.next(3); // return 4.66667 = (1 + 10 + 3) / 3
movingAverage.next(5); // return 6.0 = (10 + 3 + 5) / 3

In [ ]:
# from 
class movingavg:
    def __init__(self,size:int) -> None:
        self.size = size
        self.q=deque()
        self.total=0

    def next(self,val:int):
        q.append(val)
        total += val
        if len(q)> self.size:
            total -= q.popleft()
    
    return total/size



239. Sliding Window Maximum

You are given an array of integers nums, there is a sliding window of size k which is moving from the very left of the array to the very right. You can only see the k numbers in the window. Each time the sliding window moves right by one position.

Return the max sliding window.

 

Example 1:

Input: nums = [1,3,-1,-3,5,3,6,7], k = 3
Output: [3,3,5,5,6,7]
Explanation: 
Window position                Max
---------------               -----
[1  3  -1] -3  5  3  6  7       3
 1 [3  -1  -3] 5  3  6  7       3
 1  3 [-1  -3  5] 3  6  7       5
 1  3  -1 [-3  5  3] 6  7       5
 1  3  -1  -3 [5  3  6] 7       6
 1  3  -1  -3  5 [3  6  7]      7
Example 2:

Input: nums = [1], k = 1
Output: [1]
 

Constraints:

1 <= nums.length <= 105
-104 <= nums[i] <= 104
1 <= k <= nums.length

In [45]:
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        win = deque(nums[0:k])   # first window
        output = []

        i = 0
        while i <= len(nums) - k:
            # record max of current window
            output.append(max(win))

            # slide the window forward
            if i + k < len(nums):
                win.popleft()
                win.append(nums[i + k])

            i += 1

        return output


# Example usage
obj = Solution()
print(obj.maxSlidingWindow([1,3,-1,-3,5,3,6,7], 3))
# Output: [3, 3, 5, 5, 6, 7]


[3, 3, 5, 5, 6, 7]


the bottleneck in your current code is the repeated *max(win)* call. That’s O(k) each time, so the whole algorithm is O(n·k). For large inputs, that will time out.

Below is a better way to do the above

In [ ]:
class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        dq=deque() # Stores the index not the value
        res=[]
        for i in range(len(nums)):
            # Remove indices outside window
            if dq and dq[0]<= i-k:
                dq.popleft()
            
            # Remove elements that are smaller than the current value
            # Then add the current element
            # if the current element is not the largest then the max value is present at dq[0] 
            # and new element is added in the end of dq
            while dq and nums[dq[-1]]<= nums[i]:
                dq.pop()
            
            # adding the new value
            dq.append(i)

            # if window is complete
            # get the value at dq[0] which will be the largest because of the above while loop

            if i>=k-1:
                res.append(nums[dq[0]])
        return res
        
obj = Solution()
print(obj.maxSlidingWindow([1,3,-1,-3,5,3,6,7], 3))

[3, 3, 5, 5, 6, 7]


933. Number of Recent Calls

You have a RecentCounter class which counts the number of recent requests within a certain time frame.

Implement the RecentCounter class:

RecentCounter() Initializes the counter with zero recent requests.
int ping(int t) Adds a new request at time t, where t represents some time in milliseconds, and returns the number of requests that have happened in the inclusive range [t - 3000, t], that is, the new request plus every earlier request that is no more than 3000 milliseconds older.
It is guaranteed that every call to ping uses a strictly larger value of t than the previous call.

 

Example 1:

Input
["RecentCounter", "ping", "ping", "ping", "ping"]
[[], [1], [100], [3001], [3002]]
Output
[null, 1, 2, 3, 3]

Explanation
RecentCounter recentCounter = new RecentCounter();
recentCounter.ping(1);     // requests = [1], range is [-2999,1], return 1
recentCounter.ping(100);   // requests = [1, 100], range is [-2900,100], return 2
recentCounter.ping(3001);  // requests = [1, 100, 3001], range is [1,3001], return 3
recentCounter.ping(3002);  // requests = [1, 100, 3001, 3002], range is [2,3002], return 3
 

Constraints:

1 <= t <= 109
Each test case will call ping with strictly increasing values of t.
At most 104 calls will be made to ping.

For every req -> no of request happened at t-3000 to t


In [ ]:
from collections import deque
class RecentCounter:

    def __init__(self):
        self.q=deque()
        output=[]
        

    def ping(self, t: int) -> int:
        self.q.append(t)
        while (t-self.q[0]>3000):
            self.q.popleft()
        print(len(self.q))
        return len(self.q)

# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t)